### PR TITLE
Restructure project layout

### PR DESCRIPTION
**Pull request by pschafhalter**
- Reorganizes project into workspace containing 2 crates (erdos and python).
- Switches from `setuptools_rust` to `maturin` for managing Python distributions.
- Pin Rust version to `nightly-2021-12-04`.
- Remove benchmarks and extra dependencies.
- Update Dockerfile and CI.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -18,18 +18,21 @@ ENV SHELL /bin/bash
...
 RUN sudo apt-get -y install curl clang python3 python3-pip git
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
 ENV PATH="/home/erdos/.cargo/bin:${PATH}"
-RUN rustup default nightly
 
 # Get the erdos directory.
-RUN pip3 install setuptools setuptools-rust
 RUN mkdir -p /home/erdos/workspace
-RUN cd /home/erdos/workspace && git clone https://github.com/erdos-project/erdos.git
+RUN cd /home/erdos/workspace && git clone https://github.com/pschafhalter/erdos.git
```
Comment (ICGog) [inline Dockerfile:28]: Please switch back to erdos-project.

**Review by ICGog — COMMENTED** [review]
Looks good to me.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -47,19 +43,19 @@ jobs:
           python-version: ${{ matrix.python-version }}
       - name: Install Python dependencies
         run: |
-          python -m pip install setuptools_rust flake8 yapf flake8-quotes
+          python -m pip install maturin flake8 yapf==0.31.0 flake8-quotes
```
Comment (sukritkalra) [inline .github/workflows/ci.yaml:46]: Could we move this to a `python/requirements.txt` file?

File: .gitignore [hunk 3]
```diff
@@ -11,7 +11,7 @@ Cargo.lock
 # Python files
 *.egg-info
 __pycache__
-/env
+env
```
Comment (sukritkalra) [inline .gitignore:14]: Why was this added before?

File: Dockerfile [hunk 4]
```diff
@@ -18,18 +18,21 @@ ENV SHELL /bin/bash
...
 WORKDIR /home/erdos/workspace/erdos
 RUN cd /home/erdos/workspace/erdos
+RUN git checkout restructure
```
Comment (sukritkalra) [inline Dockerfile:31]: Remnants of your personal branch.

File: Dockerfile [hunk 5]
```diff
@@ -18,18 +18,21 @@ ENV SHELL /bin/bash
 
 SHELL ["/bin/bash", "-c"]
 
-# Instal rust.
+# Install rust.
 RUN sudo apt-get -y install curl clang python3 python3-pip git
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
 ENV PATH="/home/erdos/.cargo/bin:${PATH}"
-RUN rustup default nightly
```
Comment (sukritkalra) [inline Dockerfile:25]: Do we not need nightly anymore?

File: Dockerfile [hunk 6]
```diff
@@ -18,18 +18,21 @@ ENV SHELL /bin/bash
...
+RUN cd /home/erdos/workspace && git clone https://github.com/pschafhalter/erdos.git
 WORKDIR /home/erdos/workspace/erdos
 RUN cd /home/erdos/workspace/erdos
+RUN git checkout restructure
+
+# Install erdos.
 RUN cargo build --release
 # Install the python package
-RUN python3 python/setup.py install --user
+ENV PATH="/home/erdos/.local/bin:${PATH}"
```
Comment (sukritkalra) [inline Dockerfile:36]: You seem to be setting the `PATH` twice?

File: README.md [hunk 7]
```diff
@@ -48,34 +47,26 @@ To develop an ERDOS application in Python, simply run
...
 [install Rust](https://www.rust-lang.org/tools/install).
-Then run the following to clone the repository and build ERDOS:
+Within a [virtual environment](https://docs.python.org/3/tutorial/venv.html),
+run the following to clone the repository and build ERDOS:
 ```console
-rustup default nightly  # use nightly Rust toolchain
-git clone https://github.com/erdos-project/erdos.git && cd erdos
-python3 python/setup.py develop
+git clone https://github.com/erdos-project/erdos.git && cd erdos/python
+pip3 install maturin
```
Comment (sukritkalra) [inline README.md:54]: We should maintain a `requirements.txt`.

**Review by sukritkalra — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (pschafhalter) [reply .github/workflows/ci.yaml:46]: Packages like yapf and flake8 are used in CI to validate formatting, and aren't necessary to run ERDOS. Also, I don't think we need a `requirements.txt`, since the `pyproject.toml` specifies the dependencies.

File: .gitignore (hunk 3 above)
Comment (pschafhalter) [reply .gitignore:14]: https://github.com/github/gitignore/blob/752b7678db497edb377fa982ded0c70035393794/Python.gitignore#L114-L121

File: Dockerfile (hunk 5 above)
Comment (pschafhalter) [reply Dockerfile:25]: Cargo reads the `rust-toolchain.toml` which ensures that ERDOS will be built with Rust nightly-2021-12-04.

File: Dockerfile (hunk 6 above)
Comment (pschafhalter) [reply Dockerfile:36]: The first one sets up Cargo, this one sets up programs installed via pip (e.g. maturin).
I can combine them into 1 line, but I thought updating the `PATH` down here with the rest of the python logic made more sense.

File: README.md (hunk 7 above)
Comment (pschafhalter) [reply README.md:54]: I disagree -- the `pyproject.toml` already specifies the package dependencies. With newer pip versions, you can even run `pip install -e .` from the `python/` dir and it'll install without having installed `maturin` ahead of time.

Comment (sukritkalra) [reply README.md:54]: Did not know that! Okay fair enough, I'll resolve this and the corresponding conversation then. Thanks!

File: .gitignore (hunk 3 above)
Comment (sukritkalra) [reply .gitignore:14]: We should copy all of them since not everyone uses the `env` name. I prefer to use `venv`.

File: Dockerfile (hunk 6 above)
Comment (sukritkalra) [reply Dockerfile:35]: Apologies, I didn't see the append of the `$PATH` at the end. This should work fine!

**Review by sukritkalra — APPROVED** [review]
LGTM! Thanks for addressing the minor comments!
